File: app/src/main/java/com/mei/http/MainActivity.java
package com.mei.http;

import com.mei.http.net.Call;
import com.mei.http.net.CallBack;
import com.mei.http.net.HttpClient;
import com.mei.http.net.Request;
import com.mei.http.net.RequestBody;
import com.mei.http.net.Response;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.util.Log;
import android.view.View;

public class MainActivity extends AppCompatActivity {

    HttpClient client;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        client = new HttpClient.Builder().retry(3).build();
    }

    public void get(View view) throws IllegalAccessException {
        Request request = new Request.Builder()
                .url("http://www.kuaidi100.com/query?type=yuantong&postid=222222222")
                .build();
        Call call = client.newCall(request);
        call.enqueue(new CallBack() {
            @Override
            public void onFailure(Call call, Throwable throwable) {
                throwable.printStackTrace();
            }

            @Override
            public void onResponse(Call call, Response response) {
                Log.e("响应体", response.getBody());

            }
        });
    }

    public void post(View view) throws IllegalAccessException {
        RequestBody body = new RequestBody()
                .add("city", "长沙")
                .add("key", "13cb58f5884f9749287abbead9c658f2");
        Request request = new Request.Builder().url("http://restapi.amap" +
                ".com/v3/weather/weatherInfo").post(body).build();
        client.newCall(request).enqueue(new CallBack() {
            @Override
            public void onFailure(Call call, Throwable throwable) {
                throwable.printStackTrace();
            }

            @Override
            public void onResponse(Call call, Response response) {
                Log.e("响应体", response.getBody());
            }
        });
    }
}

File: retrofit/src/main/java/com/mei/retrofit/MainActivity.kt
package com.mei.retrofit

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.lifecycle.lifecycleScope
import com.mei.retrofit.databinding.ActivityMainBinding
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.ResponseBody
import retrofit2.*
import java.io.IOException
import java.lang.reflect.Type

class MainActivity : AppCompatActivity() {

    private val mBinding: ActivityMainBinding by lazy {
        ActivityMainBinding.inflate(layoutInflater)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(mBinding.root)
    }

    fun baidu(view: android.view.View) {
        val retrofit = Retrofit.Builder()
            .baseUrl("https://www.baidu.com")
            .addConverterFactory(object : Converter.Factory() {
                override fun responseBodyConverter(
                    type: Type,
                    annotations: Array<out Annotation>,
                    retrofit: Retrofit
                ): Converter<ResponseBody, *> {
                    return object : Converter<ResponseBody, String> {
                        override fun convert(value: ResponseBody): String? {
                            return value.string()
                        }
                    }
                }
            }).build()
        val service = retrofit.create(ApiService::class.java)

        val call: Call<String> = service.baidu()
        call.enqueue(object : Callback<String> {
            override fun onResponse(call: Call<String>, response: Response<String>) {
                println("当前线程：${Thread.currentThread()}")
                println("请求成功：${response.body()}")
                mBinding.tvMsg.text = response.body()
            }

            override fun onFailure(call: Call<String>, t: Throwable) {
                println("请求失败")
                t.printStackTrace()
            }
        })
    }
}